pkg/repository/prisma/tenant_buffer.go

```diff
@@ -0,0 +1,121 @@
+package prisma
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog"
+)
+
+// This is a wrapper around the IngestBuf to manage multiple tenants
+// An example would be T is eventOps and U is *dbsqlc.Event
+
+type TenantBufferManager[T any, U any] struct {
+	tenants     sync.Map
+	l           *zerolog.Logger
+	defaultOpts IngestBufOpts[T, U]
+}
+
+type TenantBufManagerOpts[T any, U any] struct {
+	OutputFunc func(ctx context.Context, items []T) ([]U, error)
+	SizeFunc   func(T) int
+	l          *zerolog.Logger
+}
+
+// Create a new TenantBufferManager with generic types T for input and U for output
+func NewTenantBufManager[T any, U any](opts TenantBufManagerOpts[T, U]) (*TenantBufferManager[T, U], error) {
+
+	if opts.OutputFunc == nil {
+		return nil, fmt.Errorf("output function is required")
+	}
+
+	if opts.SizeFunc == nil {
+		return nil, fmt.Errorf("size function is required")
+	}
+	if opts.l == nil {
+		return nil, fmt.Errorf("logger is required")
+	}
+
+	megabyte := 1024 * 1024
+
+	defaultOpts := IngestBufOpts[T, U]{
+		maxCapacity:        1000,
+		flushPeriod:        50 * time.Millisecond,
+		maxDataSizeInQueue: 4 * megabyte,
```

Comment:
what happens if we put a single event > 4mb (or reduce this to 3 and put a 4 mb)

Reply:
A single event greater than 4MB will flush straight away. If we add a 3mb first and then a 4mb we will be over the size and we will flush.